src/restruct_poc/beams.hpp

```diff
@@ -0,0 +1,609 @@
+//------------------------------------------------------------------------------
+
+struct Beams {
+    size_t num_beams_;  // Number of beams
+    size_t num_nodes_;  // Number of nodes
+    size_t num_qps_;    // Number of quadrature points
+
+    Kokkos::View<BeamElemIndices*> elem_indices;  // View of element node and qp indices into views
+    Kokkos::View<int*> node_state_indices;        // State row index for each node
+
+    oturb::View_3 gravity;
+
+    // Node-based data
+    oturb::View_Nx7 node_x0;              // Inital position/rotation
+    oturb::View_Nx7 node_u;               // State: translation/rotation displacement
+    oturb::View_Nx6 node_u_dot;           // State: translation/rotation velocity
+    oturb::View_Nx6 node_u_ddot;          // State: translation/rotation acceleration
+    oturb::View_Nx6 node_force_app;       // Applied forces
+    oturb::View_Nx6 node_force_elastic;   // Elastic forces
+    oturb::View_Nx6 node_force_gravity;   // Gravity forces
+    oturb::View_Nx6 node_force_inertial;  // Inertial forces
+    oturb::View_Nx6 node_force_external;  // External forces
+    oturb::View_Nx6 node_residual;        // Residual forces
+    oturb::View_NxN node_iteration;       // Iteration matrix
+
+    // Quadrature point data
+    oturb::View_N qp_weight;       // Integration weights
+    oturb::View_N qp_jacobian;     // Jacobian vector
+    oturb::View_Nx6x6 qp_Mstar;    // Mass matrix in material frame
+    oturb::View_Nx6x6 qp_Cstar;    // Stiffness matrix in material frame
+    oturb::View_Nx3 qp_x0;         // Initial position
+    oturb::View_Nx3 qp_x0_prime;   // Initial position derivative
+    oturb::View_Nx4 qp_r0;         // Initial rotation
+    oturb::View_Nx3 qp_u;          // State: translation displacement
+    oturb::View_Nx3 qp_u_prime;    // State: translation displacement derivative
+    oturb::View_Nx3 qp_u_dot;      // State: translation velocity
+    oturb::View_Nx3 qp_u_ddot;     // State: translation acceleration
+    oturb::View_Nx4 qp_r;          // State: rotation
+    oturb::View_Nx4 qp_r_prime;    // State: rotation derivative
+    oturb::View_Nx3 qp_omega;      // State: angular velocity
+    oturb::View_Nx3 qp_omega_dot;  // State: position/rotation
+    oturb::View_Nx6 qp_strain;     //
+    oturb::View_Nx6 qp_Fc;         //
+    oturb::View_Nx6 qp_Fd;         //
+    oturb::View_Nx6 qp_Fi;         //
+    oturb::View_Nx6 qp_Fg;         //
+    oturb::View_Nx6 qp_Fext;       //
+    oturb::View_Nx6x6 qp_RR0;      //
+    oturb::View_Nx6x6 qp_Muu;      //
+    oturb::View_Nx6x6 qp_Cuu;      //
+    oturb::View_Nx6x6 qp_Ouu;      //
+    oturb::View_Nx6x6 qp_Puu;      //
+    oturb::View_Nx6x6 qp_Quu;      //
+    oturb::View_Nx6x6 qp_Guu;      //
+    oturb::View_Nx6x6 qp_Kuu;      //
+
+    oturb::View_NxN shape_interp;  // shape function matrix for interpolation [Nodes x QPs]
+    oturb::View_NxN shape_deriv;   // shape function matrix for derivative interp [Nodes x QPs]
```

Comment:
This will be a block diagonal matrix, in general, so the dense matrix representation will have a ton of 0 multiplications.  

With the "single number of nodes/qp for all elements" structure I discussed, we could use single "nodes-per-element X qp-per-element" matrix for all elements.  We would then need one interpolation matrix for each element type when we generalize to elements with different numbers of nodes.  

With the current structure, we will need to use a sparse matrix representation here to get good performance for interpolation.  Kokkos has one that we'll need to learn how to use for the linear system anyways...

Reply:
The comment on those views is a little misleading. It should say `[Nodes x Max Element QPs]`.  The number of columns is equal to the maximum number of quadrature points in any element. Since the number of quadrature points should be pretty similar or bounded, it didn't make sense to allocate a full `[Nodes x QPs]` for all the reasons you describe. The current code avoids the sparsity at the cost of having to perform the interpolation matrix multiplication on an element by element basis. But those matrix operations will be on fully-populated dense matrices.

Reply:
Right, but the structure as you have it is still square and will be sparse in general.  In the event that some elements have many quadrature points and some have few, we will have a lot of unused entries.  If we have one element with 2 nodes and one element with 10 nodes (and same number of quad points), the structure will look something like:
X X 0 0 0 0 0 0 0 0
X X 0 0 0 0 0 0 0 0
X X X X X X X X X X
X X X X X X X X X X
X X X X X X X X X X
...

That block of 0 elements gets really big when there are more low order elements or larger high order elements.  I get that you use an indexing strategy so that we don't have to do math on the 0 entries, but it will use a lot of memory.

Reply:
I agree that it has the potential to waste an appreciable amount of memory. However, it's unlikely that there will be a significant difference in the number of quadrature points between elements. OpenFAST caps the number of quadrature points at 7 and most elements will probably need at least 3. We could record the usage percent in the data structure and see what it's like under various use cases. I'm open to an alternative representation.